pkg/streamingpromql/operators/functions/info.go

```diff
@@ -0,0 +1,575 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package functions
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/operators/selectors"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+	"github.com/grafana/mimir/pkg/util/limiter"
+)
+
+// identifyingLabels are the labels we consider as identifying for info metrics.
+// Currently hard coded, so we don't need knowledge of individual info metrics.
+var identifyingLabels = []string{"instance", "job"}
+
+type labelsTime struct {
+	labels labels.Labels
+	time   int64
+}
+
+type InfoFunction struct {
+	Inner                    types.InstantVectorOperator
+	Info                     types.InstantVectorOperator
+	MemoryConsumptionTracker *limiter.MemoryConsumptionTracker
+
+	timeRange                types.QueryTimeRange
+	expressionPosition       posrange.PositionRange
+	enableDelayedNameRemoval bool
+
+	// function to generate signature from labels without metric name
+	sigFunctionLabelsOnly func(labels.Labels) string
+	// labels hash:function to generate signature from labels
+	sigFunctions map[string]func(labels.Labels) string
+	// timestamp:(labels only signature:array of labels)
+	sigLabelsOnlyTimestamps map[int64]map[string][]labels.Labels
+	// labels only signature:(label sets hash:array of labels)
+	labelSets map[string]map[string][]labels.Labels
+	// inner series index - (info series label sets hash: index for ordering)
+	labelSetsOrder []map[string]int
+	// inner series index - inner series labels only signature
+	innerSigLabelsOnly []string
+	// stored series results for current inner series
+	storedSeriesResults []types.InstantVectorSeriesData
+
+	nextInnerSeriesIndex  int
+	nextStoredSeriesIndex int
+}
+
+func NewInfoFunction(
+	inner types.InstantVectorOperator,
+	info types.InstantVectorOperator,
+	memoryConsumptionTracker *limiter.MemoryConsumptionTracker,
+	timeRange types.QueryTimeRange,
+	expressionPosition posrange.PositionRange,
+	enableDelayedNameRemoval bool,
+) *InfoFunction {
+	return &InfoFunction{
+		Inner:                    inner,
+		Info:                     info,
+		MemoryConsumptionTracker: memoryConsumptionTracker,
+
+		timeRange:                timeRange,
+		expressionPosition:       expressionPosition,
+		enableDelayedNameRemoval: enableDelayedNameRemoval,
+	}
+}
+
+func (f *InfoFunction) SeriesMetadata(ctx context.Context, matchers types.Matchers) ([]types.SeriesMetadata, error) {
+	ivs, ok := f.Info.(*selectors.InstantVectorSelector)
+	if !ok {
+		return nil, fmt.Errorf("info function 2nd argument is not an instant vector selector")
+	}
+	// Override float values to reflect original timestamps.
+	ivs.ReturnSampleTimestampsPreserveHistograms = true
```

Comment:
If we require `Info` to be an `InstantVectorSelector`, then we should enforce this by changing the type of the parameter to `NewInfoFunction`.